File: Cpp/Beginner/03_BinarySearch/boj_1654.cpp
//
// Created by seungsu on 22. 11. 19.
//

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

static long long solve(vector<long long>& arr, int n) {
    long long left = 1;
    long long right = *max_element(arr.begin(), arr.end());
    long long ans = 0;

    while (left <= right) {
        long long mid = (left + right) / 2;
        long long tmp = 0;

        for (long long v : arr)
            tmp += v / mid;

        if (tmp < n) right = mid - 1;
        else {
            left = mid + 1;
            ans = max(ans, mid);
        }
    }

    return ans;
}

void boj_1654() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int k, n;
    cin >> k >> n;

    vector<long long> arr(k);
    for (int i = 0; i < k; i++)
        cin >> arr[i];

    cout << solve(arr, n);
}

File: Python/beginner/03_BinarySearch/boj_1654.py
import sys

input = sys.stdin.readline 

# n개를 만들 수 있는 랜선의 최대 길이를 구하라
def solve(arr: list[int], n: int) -> int:
    left = 1
    right = max(arr)
    ans = 0

    while left <= right:
        mid = (left + right) // 2
        tmp = 0

        for val in arr:
            tmp += val // mid
        
        if tmp < n:
            right = mid - 1
        else:
            left = mid + 1
            ans = max(ans, mid)     # n개를 만들 수 있을 때 답을 더 큰 값으로 계속 갱신

    return ans

if __name__ == '__main__':
    k, n = map(int, input().split())
    arr = [int(input()) for _ in range(k)]

    print(solve(arr, n))